packages/cozy-harvest-lib/src/helpers/konnectorBlock.js

```diff
@@ -0,0 +1,227 @@
+  buildIsAccountConnectedQuery,
+  buildIsInErrorTriggersQuery,
+  buildTriggersByIdQuery
+} from './queries'
+import { hasNewVersionAvailable } from './konnectors'
+
+const formatKonnector = ({
+  client,
+  konnector,
+  universalLink,
+  iconStatus,
+  message
+}) => {
+  const link = generateUniversalLink({
+    cozyUrl: client.getStackClient().uri,
+    slug: universalLink.slug,
+    subDomainType: client.getInstanceOptions().cozySubdomainType,
+    nativePath: universalLink.nativePath
+  })
+
+  return {
+    name: konnector.name,
+    link,
+    vendorLink: konnector.vendor_link && {
+      component: 'a',
+      href: konnector.vendor_link,
+      target: '_blank'
+    },
+    iconStatus,
+    message
+  }
+}
+
+const getDataFromState = async (client, query) => {
+  try {
+    await client.query(query.definition, query.options)
+    const { data } = client.getQueryFromState(query.options.as)
+    return data[0]
+  } catch (error) {
+    throw error
+  }
+}
+
+const getKonnector = (client, query) => getDataFromState(client, query)
+
+const isAccountConnected = async (client, sourceAccount) => {
+  const query = buildIsAccountConnectedQuery(sourceAccount)
+  const data = await getDataFromState(client, query)
+  return !!data
+}
+
+// Only konnectors from the registry have a maintenance status
+const isInMaintenance = async (client, slug) => {
+  const konnector = await konnectorBlock.getKonnector(
+    client,
+    buildKonnectorQuery(slug, 'registry')
+  )
+  return get(konnector, 'maintenance_activated', false)
+}
+
+const isInError = async ({ client, slug, sourceAccount }) => {
+  const query = buildIsInErrorTriggersQuery(slug, sourceAccount)
+  const data = await getDataFromState(client, query)
+  const triggerId = get(data, 'id')
+
+  const triggersByIdQuery = buildTriggersByIdQuery(triggerId)
+  const trigger = await getDataFromState(client, triggersByIdQuery)
+
+  const inError = get(trigger, 'current_state.status') === 'errored'
+
+  if (!inError) {
+    return { error: null }
+  }
+
+  const message = get(trigger, 'current_state.last_error')
+
+  const isActionable = triggers.hasActionableError(trigger)
+
+  return { error: { message, isActionable } }
+}
+
+export const getKonnectorData = async ({ client, t, slug, sourceAccount }) => {
+  try {
+    const konnector = await konnectorBlock.getKonnector(
+      client,
+      buildKonnectorQuery(slug, 'stack')
+    )
+
+    const accountConnected = await konnectorBlock.isAccountConnected(
+      client,
+      sourceAccount
+    )
+
+    if (!accountConnected) {
+      return formatKonnector({
+        client,
+        konnector: konnector.attributes,
+        universalLink: {
+          slug: 'home',
+          nativePath: `connected/${slug}/new`
+        },
+        iconStatus: 'disabled',
+        message: {
+          text: t('konnectorBlock.disconnected')
+        }
+      })
+    }
+
+    const inMaintenance = await konnectorBlock.isInMaintenance(client, slug)
```

Comment:
At this point, why isInMaintenance needs to be passed the client, couldn't it check the value of the maintenance flag directly from the `konnector` variable ?

Reply:
Tu parles du `konnector` récupéré plus haut via la stack ? Si c'est ça, ce `konnector` ne porte pas d'information de maintenance, l'info n'est présente que suite à un fetch de la registry. Donc on doit passer `client` à une fonction afin de faire un `client.query`.
Je sais pas si j'ai bien bien compris ta question (et donc apporté une réponse assez pertinente) ?

Reply:
Ah oui c'est vrai OK je comprends mieux. Dans ce cas là à mon avis ca aurait plus de sens d'avoir un 

```
const qDef = Q('io.cozy.konnectors').getById(slug)
qDef.source = 'registry' // je crois que c'est comme ça qu'on l'a fait ?
const { registryKonnector } = await client.query(qDef, { fetchPolicy: defaultFetchPolicy })
const inmaintenance = konnectorModel.isInMaintenance(registryKonnector)
```

A mon avis il ne faut pas mettre d'async sur les fonctions isInMaitenance, isInError, il faut remonter le data fetching et pas le mettre dans des fonctions helpers, sinon je trouve ca difficile d'auditer qu'est ce qui est fait et quand.

Reply:
A mon avis ca serait plus clair si séparait le calcul du "statut" du connecteur des informations à présenter.


```
const statusToFormatOptions  =  {
  inMaintenance: konnector => ({
    universalLink: {
      slug: 'home',
      nativePath: 'connected/${slug}'
    }),
    iconStatus: 'disabled',
    message: { ...}
  }),
  noAccountConnected: (...) => {...}
}


const fetchKonnectorStatus = () => {
  // does all the fetching and logic to determine in which case we are
}

const fetchKonnectorData = (konnector) => {
  const status = await fetchKonnectorStatus(konnector)
  return statusToFormatOptions[status](konnector)
}
```

De cette manière là je trouve qu'on verrait mieux les cas possibles. Et on découple le statut (model) de la présentation.

Reply:
L'idée c'était d'avoir un `fetchKonnectorData` assez simple à lire, mais du coup oui ça masque un peu ce qui se passe sous le capot donc une première lecture ne suffit pas à comprendre concrètement ce qui se passe... perso je préfère ça permet de se concentrer d'abord sur l'algo, puis ensuite sur le "comment", mais bon c'est peut-être pas la meilleure manière de faire...

Pourquoi ne pas mettre d'async sur les fonctions isInMaitenance, isInError ?
Par contre dans ton exemple il faudrait rajouter le `getQueryFromState` car `client.query` ne retourne rien s'il y a du cache.

L'idée de faire des fonctions autonomes avec le fetch dedans était aussi pour simplifier les tests, suffit de spy toute la fonction et de retourner une valeur.


Ton deuxième exemple me plaît bien c'est ce vers quoi je voulais tendre, sans y être arrivé. Je vais creuser cette piste.